### PR TITLE
[SPARK-23081][PYTHON]Add colRegex API to PySpark

### PR DESCRIPTION
**Pull request by huaxingao**
## What changes were proposed in this pull request?

Add colRegex API to PySpark

## How was this patch tested?

add a test in sql/tests.py

### CODE REVIEW

File: python/pyspark/sql/dataframe.py [hunk 1]
```diff
@@ -1881,6 +1881,15 @@ def toDF(self, *cols):
         jdf = self._jdf.toDF(self._jseq(cols))
         return DataFrame(jdf, self.sql_ctx)
 
+    @since(2.3)
```
Comment (HyukjinKwon) [inline python/pyspark/sql/dataframe.py]: I think this should be 2.4.

File: python/pyspark/sql/dataframe.py [hunk 2]
```diff
@@ -1881,6 +1881,15 @@ def toDF(self, *cols):
...
+    def colRegex(self, colName):
+        """
+        Selects column based on the column name specified as a regex and return it
+        as :class:`Column`.
```
Comment (HyukjinKwon) [inline python/pyspark/sql/dataframe.py]: Shall we add a doctest and `:param ` too while we are here?

File: python/pyspark/sql/dataframe.py [hunk 3]
```diff
@@ -1881,6 +1881,15 @@ def toDF(self, *cols):
...
+        """
+        jc = self._jdf.colRegex(colName)
```
Comment (HyukjinKwon) [inline python/pyspark/sql/dataframe.py]: Could we add a type check here too?

Comment (huaxingao) [reply python/pyspark/sql/dataframe.py]: @HyukjinKwon Thank you very much for your comments. I will submit changes soon.

File: python/pyspark/sql/dataframe.py [hunk 4]
```diff
@@ -1881,6 +1881,28 @@ def toDF(self, *cols):
         jdf = self._jdf.toDF(self._jseq(cols))
         return DataFrame(jdf, self.sql_ctx)
 
+    @since(2.4)
```
Comment (HyukjinKwon) [inline python/pyspark/sql/dataframe.py]: Could we put this API between `def columns(self):` and `def alias(self, alias):`?

File: python/pyspark/sql/tests.py [hunk 5]
```diff
@@ -2855,6 +2855,10 @@ def test_create_dataframe_from_old_pandas(self):
             with self.assertRaisesRegexp(ImportError, 'Pandas >= .* must be installed'):
                 self.spark.createDataFrame(pdf)
 
+    def test_colRegex(self):
+        df = self.spark.createDataFrame([("a", 1), ("b", 2), ("c",  3)])
+        self.assertEqual(df.select(df.colRegex("`(_1)?+.+`")).collect(), df.select("_2").collect())
```
Comment (HyukjinKwon) [inline python/pyspark/sql/tests.py]: I think this is actually being tested in doctest. Seems we can remove out.

File: python/pyspark/sql/dataframe.py [hunk 6]
```diff
@@ -1881,6 +1881,28 @@ def toDF(self, *cols):
...
+        >>> df.select(df.colRegex("`(_1)?+.+`")).show()
+        +---+
+        | _2|
+        +---+
+        |  1|
+        |  2|
+        |  3|
+        +---+
+        """
+        assert isinstance(colName, basestring), "colName should be a string"
```
Comment (HyukjinKwon) [inline python/pyspark/sql/dataframe.py]: I think `TypeError` with an if could be more correct.

**Review by HyukjinKwon — APPROVED** [review]
LGTM otherwise.

File: python/pyspark/sql/tests.py (hunk 5 above)
Comment (huaxingao) [reply python/pyspark/sql/tests.py]: @HyukjinKwon Thanks! I will make the changes.

File: python/pyspark/sql/dataframe.py [hunk 7]
```diff
@@ -819,6 +819,29 @@ def columns(self):
...
+    @since(2.4)
+    def colRegex(self, colName):
+        """
+        Selects column based on the column name specified as a regex and return it
+        as :class:`Column`.
+
+        :param colName: string, column name specified as a regex.
+
+        >>> df = spark.createDataFrame([("a", 1), ("b", 2), ("c",  3)])
+        >>> df.select(df.colRegex("`(_1)?+.+`")).show()
```
Comment (felixcheung) [inline python/pyspark/sql/dataframe.py]: nit: perhaps a bit obscure to pick the default column name of `_1`?
how about we name the columns in the line above?

**Review by felixcheung — APPROVED** [review]

File: python/pyspark/sql/dataframe.py (hunk 7 above)
Comment (huaxingao) [reply python/pyspark/sql/dataframe.py]: @felixcheung Thanks for your comment! I will make changes.

File: python/pyspark/sql/dataframe.py [hunk 8]
```diff
@@ -819,6 +819,29 @@ def columns(self):
         """
         return [f.name for f in self.schema.fields]
 
+    @since(2.4)
+    def colRegex(self, colName):
+        """
+        Selects column based on the column name specified as a regex and return it
```
Comment (gatorsmile) [inline python/pyspark/sql/dataframe.py]: Nit: -> `returns`

File: python/pyspark/sql/dataframe.py [hunk 9]
```diff
@@ -819,6 +819,29 @@ def columns(self):
         """
         return [f.name for f in self.schema.fields]
 
+    @since(2.4)
```
Comment (gatorsmile) [inline python/pyspark/sql/dataframe.py]: -> `2.3`

**Comment by gatorsmile** [comment]
Since our Spark 2.3 RC2 will fail, we can target it to 2.3

File: python/pyspark/sql/dataframe.py (hunk 8 above)
Comment (gatorsmile) [reply python/pyspark/sql/dataframe.py]: Unfortunately, we have the same issue in Dataset.colRegex. Please also correct that too.

**Comment by gatorsmile** [comment]
LGTM except the above two comments.

File: python/pyspark/sql/dataframe.py (hunk 8 above)
Comment (huaxingao) [reply python/pyspark/sql/dataframe.py]: @gatorsmile Thanks for your comments. I will make the changes.

**Comment by felixcheung** [comment]
Awesome
LGTM pending test passes

**Comment by HyukjinKwon** [comment]
Merged to master and branch-2.3.

**Comment by huaxingao** [comment]
Thank you all for your help! @HyukjinKwon @gatorsmile @felixcheung

**Pull request closed without merging**